File: cloudcode-backend/backend/services/analysis_service.py
# cloudcode_backend/services/analysis_service.py
import subprocess
import json
import os
from datetime import datetime
from ..utils.temp_dir import temporary_dir

def run_command(command, working_dir):
    """Executes a shell command and returns its output."""
    result = subprocess.run(command, cwd=working_dir, capture_output=True, text=True, check=True)
    return result.stdout.strip()

def analyze_git_repository(repo_url: str) -> dict:
    """
    Clones a Git repository and performs a comprehensive analysis.
    """
    with temporary_dir() as temp_dir:
        # 1. Clone the repository
        subprocess.run(['git', 'clone', '--depth=1', repo_url, '.'], cwd=temp_dir, check=True)

        # 2. Get project name
        project_name = os.path.basename(repo_url).replace('.git', '')

        # 3. Run cloc for language breakdown
        cloc_json_output = run_command(['cloc', '--json', '.'], temp_dir)
        cloc_data = json.loads(cloc_json_output)
        
        # Remove the header and summary for cleaner language data
        cloc_data.pop('header', None)
        summary = cloc_data.pop('SUM', {})

        # 4. Gather Git metadata
        last_commit_date = run_command(['git', 'log', '-1', '--format=%cI'], temp_dir)
        total_commits = run_command(['git', 'rev-list', '--count', 'HEAD'], temp_dir)
        contributor_count = len(run_command(['git', 'shortlog', '-s', '-n', '--all'], temp_dir).splitlines())
        active_branches = len(run_command(['git', 'branch', '-r'], temp_dir).splitlines())

        analysis_report = {
            "projectName": project_name,
            "repositoryUrl": repo_url,
            "lastAnalysisDate": datetime.now().isoformat(),
            "identification": {
                "languageBreakdown": cloc_data,
                "totalFiles": summary.get("nFiles", 0),
                "totalLinesOfCode": summary.get("code", 0),
                "dependencyCount": "N/A" # Placeholder
            },
            "activity": {
                "lastCommitDate": last_commit_date,
                "totalCommits": int(total_commits),
                "contributorCount": contributor_count,
                "commitFrequency": "N/A", # Placeholder
                "activeBranches": active_branches
            },
            "qualityHealth": { # Placeholders for future implementation
                "technicalDebtRatio": "N/A",
                "codeIssues": {"critical": "N/A", "major": "N/A", "minor": "N/A"},
                "codeCoverage": "N/A",
                "cyclomaticComplexity": "N/A",
                "duplicatedLines": "N/A"
            }
        }
        return analysis_report

File: cloudcode/src/services/AnalysisService.ts
// src/services/AnalysisService.ts

import * as vscode from 'vscode';
import * as path from 'path';
import { ProjectAnalysis } from '../core/types';
import { performLocalAnalysis } from '../utilities/fileSystemUtils';
import { ApiService } from './ApiService';

/**
 * Creates a project analysis report for a local-only folder.
 * @param projectRoot The root path of the project.
 * @returns A ProjectAnalysis object with local data.
 */
export async function analyzeLocalFolder(projectRoot: string): Promise<ProjectAnalysis> {
    const analysisResult = await performLocalAnalysis(projectRoot);
    const summary = analysisResult.SUM;
    delete analysisResult.SUM;

    // We construct a ProjectAnalysis object, filling in what we can from local data.
    return {
        projectName: path.basename(projectRoot) + " (Local)",
        repositoryUrl: "Local Folder",
        lastAnalysisDate: new Date().toISOString(),
        identification: {
            languageBreakdown: analysisResult,
            totalFiles: summary.nFiles,
            totalLinesOfCode: summary.code,
            dependencyCount: "N/A"
        },
        activity: {
            lastCommitDate: "N/A",
            totalCommits: "N/A",
            contributorCount: "N/A",
            commitFrequency: "N/A",
            activeBranches: "N/A"
        },
        qualityHealth: {
            technicalDebtRatio: "N/A",
            codeIssues: { critical: "N/A", major: "N/A", minor: "N/A" },
            codeCoverage: "N/A",
            cyclomaticComplexity: "N/A",
            duplicatedLines: "N/A"
        }
    };
}

/**
 * Orchestrates a remote repository analysis via the backend.
 * @param repoUrl The URL of the Git repository.
 * @param progress A VS Code progress object to report updates to the user.
 * @returns A promise that resolves to the full project analysis from the backend.
 */
export async function analyzeRemoteRepo(repoUrl: string, progress: vscode.Progress<{ message?: string; increment?: number }>): Promise<ProjectAnalysis> {
    progress.report({ increment: 30, message: "Connecting to analysis backend..." });
    const result = await ApiService.analyzeRemoteRepo(repoUrl);
    progress.report({ increment: 80, message: "Processing backend data..." });
    return result;
}